File: 1373-max-sum-BST/solution.go
package main

import (
	"gopractice/datastructure/Tree"
	"math"
)

type TreeNode = Tree.TreeNode

type subTree struct {
	IsBST  bool
	MinVal int
	MaxVal int
	Sum    int
}

var maxSum int

func maxSumBST(root *TreeNode) int {
	maxSum = 0
	dfs(root)
	return maxSum
}

func maxInt(max int, cur int) int {
	if cur > max {
		return cur
	}
	return max
}

func minInt(min int, cur int) int {
	if cur < min {
		return cur
	}
	return min
}

func dfs(root *TreeNode) *subTree {
	if root == nil {
		return &subTree{true, math.MaxInt, math.MinInt, 0}
	}
	left := dfs(root.Left)
	right := dfs(root.Right)
	if left.IsBST && right.IsBST && root.Val.(int) > left.MaxVal && root.Val.(int) < right.MinVal {
		curSum := left.Sum + right.Sum + root.Val.(int)
		maxSum = maxInt(maxSum, curSum)
		lower := minInt(root.Val.(int), left.MinVal)
		upper := maxInt(root.Val.(int), right.MaxVal)
		return &subTree{true, lower, upper, curSum}
	}
	return &subTree{false, 0, 0, 0}
}

/*
自己寫不出來, 看官解後自行理解
官解:
https://leetcode.cn/problems/maximum-sum-bst-in-binary-tree/solutions/2276045/er-cha-sou-suo-zi-shu-de-zui-da-jian-zhi-lii4/

必需滿足:
左子樹為BST
右子樹為BST
左子樹最大 < 當前值 < 右子樹最小

lower代表下界--->找左子樹中的最大
upper代表上界--->找右子樹中的最小

相當於由後往前推倒整顆BST的上下界
*/

File: 1373-max-sum-BST/solution.js
const TreeNode = require('../datastructure/Tree/TreeNode');
const TreeDebugger = require('../datastructure/Tree/TreeDebugger')
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxSumBST = function(root) {
    let [curSum, maxSum] = [0, 0]
    function dfs(root){
        if (root === null)
            return 0;
        let leftSum = dfs(root.left)
        let rightSum = dfs(root.right)
        const isBST = valid(root, -Infinity, Infinity)
        if (isBST){
            curSum = leftSum + rightSum + root.val
            maxSum = Math.max(curSum, maxSum)
        }
        else
            return 0

        return curSum
    }
    function valid(root, lower, upper){
        if (root === null)
            return true
        if (root.val <= lower || root.val >= upper)
            return false
        let isLeftBST = valid(root.left, lower, root.val)
        let isRightBST = valid(root.right, root.val, upper)
        return isLeftBST && isRightBST
    }
    dfs(root)
    return maxSum
};
/*
看過每個節點為root時的最大鍵值和, 回傳統計到的最大值

檢查以當前節點為root, 它是不是一顆BST
是 計算curSum
不是 回傳0

細節:
寫一個valid, 並確保只要這樣呼叫, 就能判定 root 是否為BST
const isBST = valid(root, -Infinity, Infinity)

valid寫法直接拿這題來用
098-is-valid-BST/solution.js

這版拿去提交會超時, 先看完improve.md
1373-max-sum-BST/improve.md
再看:
1373-max-sum-BST/solution.go
*/